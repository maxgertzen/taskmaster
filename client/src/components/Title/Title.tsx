import { Theme } from '@emotion/react';
import { createElement, FC, ReactNode } from 'react';

import { Headings } from '../../types/shared';

interface TitleProps {
  variant: Headings;
  color?: Theme['colors'];
  children: ReactNode;
}

export const Title: FC<TitleProps> = ({ variant, color, children }) => {
  return createElement(variant, { css: { color } }, children);
};
