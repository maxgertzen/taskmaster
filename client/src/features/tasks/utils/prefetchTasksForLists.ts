import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
