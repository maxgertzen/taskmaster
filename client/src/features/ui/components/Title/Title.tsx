import { Theme } from '@emotion/react';
import { createElement, FC, ReactNode } from 'react';

import { Headings } from '@/shared/types/ui';

interface TitleProps {
  variant: Headings;
  color?: Theme['colors'];
  children: ReactNode;
}

export const Title: FC<TitleProps> = ({ variant, color, children }) => {
  return createElement(variant, { css: { color } }, children);
};
