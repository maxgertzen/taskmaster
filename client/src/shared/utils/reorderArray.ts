export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
