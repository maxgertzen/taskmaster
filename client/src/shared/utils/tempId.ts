export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
