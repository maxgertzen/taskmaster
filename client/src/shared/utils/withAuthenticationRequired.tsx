import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
