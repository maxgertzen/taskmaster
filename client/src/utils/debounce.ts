export function debounce<T extends (...args: Parameters<T>) => ReturnType<T>>(
  func: T,
  delay: number
): ((...args: Parameters<T>) => void) & { cancel: () => void } {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  const debouncedFunction = (...args: Parameters<T>) => {
    if (timeoutId) clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), delay);
  };

  debouncedFunction.cancel = () => {
    if (timeoutId) clearTimeout(timeoutId);
  };

  return debouncedFunction;
}
