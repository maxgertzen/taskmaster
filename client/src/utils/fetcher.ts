export type FetcherConfig<K> = {
  url: string;
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  body?: K;
  timeout?: number;
  headers?: HeadersInit;
  urlSearchParams?: string;
};

export const fetcher = async <T, K = undefined>(
  config: FetcherConfig<K>
): Promise<T> => {
  try {
    const {
      method = 'GET',
      body = {},
      headers = {},
      timeout = 5000,
      urlSearchParams,
    } = config;
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    const options: RequestInit = {
      method,
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
      ...(method !== 'GET' && { body: JSON.stringify(body) }),
      signal: controller.signal,
    };

    if (urlSearchParams) {
      config.url += `?${urlSearchParams}`;
    }

    const response = await fetch(config.url, options);
    clearTimeout(timeoutId);

    if (!response.ok) {
      throw new Error(`Error: ${response.status} ${response.statusText}`);
    }

    return await response.json();
  } catch (error) {
    console.error('Failed to fetch data:', error);
    throw error;
  }
};
