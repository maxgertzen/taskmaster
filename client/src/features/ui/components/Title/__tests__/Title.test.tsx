import { describe, it, expect } from 'vitest';

import { render } from '@/tests/utils';

import { Title } from '../Title';

describe('Title', () => {
  it.each([
    ['h1', 'Test Title'],
    ['h2', 'Subtitle'],
    ['h3', 'Small Title'],
  ])('renders %s heading correctly', (variant, text) => {
    const { container } = render(
      <Title variant={variant as 'h1' | 'h2' | 'h3'}>{text}</Title>
    );
    expect(container).toMatchSnapshot();
  });

  it('applies custom color when provided', () => {
    const { container } = render(
      <Title variant='h1' color='red'>
        Colored Title
      </Title>
    );
    expect(container).toMatchSnapshot();
  });
});
