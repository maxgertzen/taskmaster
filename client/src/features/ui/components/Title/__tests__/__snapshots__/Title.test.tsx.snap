// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Title > applies custom color when provided 1`] = `
<div>
  <h1
    css="[object Object]"
  >
    Colored Title
  </h1>
</div>
`;

exports[`Title > renders h1 heading correctly 1`] = `
<div>
  <h1
    css="[object Object]"
  >
    Test Title
  </h1>
</div>
`;

exports[`Title > renders h2 heading correctly 1`] = `
<div>
  <h2
    css="[object Object]"
  >
    Subtitle
  </h2>
</div>
`;

exports[`Title > renders h3 heading correctly 1`] = `
<div>
  <h3
    css="[object Object]"
  >
    Small Title
  </h3>
</div>
`;
